### ISSUE TITLE
[Proposal] Registration API

### ISSUE THREAD

**Issue by oscarotero** [webcomponents-cg/community-protocols#52]
Hi. I'd like to propose a generic API to register custom elements easier, and allow to customize the tag name if it's needed. I got the inspiration [from this article](https://www.mayank.co/blog/defining-custom-elements/) from @mayank99.

The code:

```js
class MyComponent extends HTMLElement {
  static tagName = "my-component";

  static register(tagName = this.tagName) {
    customElements.define(tagName, this);
  }
}
```

This would allow to register the element in different ways:

```js
import MyComponent from "./my-component.js";

MyComponent.register(); // Register as `<my-component>`.
```

```js
import MyComponent from "./my-component.js";

MyComponent.register("your-component"); // Register as `<your-component>`.
```

```js
import MyComponent from "./my-component.js";

MyComponent.tagName = "your-component";
MyComponent.register(); // Register as `<your-component>`.
```

**Comment by thepassle** [comment]
What is the benefit of this over just doing `customElements.define`?

**Comment by Westbrook** [comment]
@oscarotero thanks for starting this conversation!!

The "Goals" section proposed in @matthewp's protocol proposal template will be a useful piece of work to smooth the discussion around any proposal. Would be great to start there (even if it's just one or two to get the conversation started) in issues opened as well, so we can all be sure to be on the same page when discussing something like this.

We will be discussing the template ratification at next week's [Web Components Community Group meeting](https://github.com/w3c/webcomponents-cg/issues/76), feel free to join in the conversation.

**Comment by thepassle** [comment]
Shouldnt an element that uses other elements internally via scoped registries just define them by default? If you forget calling .define, the element doesnt render anything? Why not do it automatically?

**Comment by keithamus** [comment]
You mean to move it into the constructor or something? So the code in my comment would instead be:

```js
let registry = new CustomElementRegistry()
class MyComponent extends HTMLElement {
  constructor() {
    super()
    if (!registry.get('my-dependent')) registry.define('my-dependent', MyDependent)
    if (!registry.get('my-other')) registry.define('my-other', MyOtherDependent)
  }
}
```

**Comment by thepassle** [comment]
Yeah, or something like we do in scoped-elements: https://github.com/open-wc/open-wc/blob/master/packages/scoped-elements/html-element.js

im not sure what the benefit of requiring a consumer of your element to manually call .define before being able to use the element is?

I do like a static property to suggest a default tagName (even though its just that; a suggestion, because a consumer of your class may register it under a different name) I also do this in generic-components, I just dont think a lot of abstractions over customElements.define actually provide anything over… just calling customElements.define.

**Comment by trusktr** [comment]
I think scoped registries solve all use cases listed so far. The only difficulty is getting all library authors to not automatically define their elements. And what about existing libraries that authors don't want to update?

**Comment by mattlucock** [comment]
I agree that a custom element shouldn't force a user to have that custom element defined by a particular name. Of course, a big benefit of the custom element defining itself is that a user who doesn't care about the name and just wants to use the default name can do so. I like the idea of class having a static property that contains an author-specified default name that a user can use if they don't want to fuss; that seems like quite a good idea, actually.

But I strongly agree with @thepassle that abstracting `customElements.define` like this is a mistake. Also, I think the idea of this tag name being mutable (as opposed to an author-specified constant) is a mistake, since it implies that an entire custom element class is only allowed to be registered once under one particular name, which isn't a constraint that custom elements normally have (and in general I don't think the benefit of this has been justified).

The biggest problem I have with this protocol proposal is: I don't think this is a protocol. This doesn't relate to interaction between different parts of a system, and it's unclear how this can be interoperable. This proposal cannot achieve "a consistent way to register web components" for "all components", since in general, most components will not support this. 

A component author could simply choose to expose this or some variation of it as part of a component's public API, and indeed, the only way you could use this is if the component author chose to make it part of the public API—in which case, this isn't some 'standard' thing; it's just that component's particular API. I don't think this is a protocol.

And indeed, we already have a consistent way to register web components, that all components support: it's `customElements.define`.

**Comment by mattlucock** [comment]
I was mistaken in my previous comment; I got confused and thought you could register a custom element twice under different names, but it's fairly intuitive that you can't, since if you instantiated a custom element directly, but it has multiple names, what the tag name would be for that instance is indeterminate.

Given that, I don't think it's obvious that user-customized tag names actually are a good idea. If two different implementations try and define the same custom element on the same page under the different names, the one that goes second will fail, which is unpredictable since in general the ordering is arbitrary. Of course, scoped registries would theoretically solve this problem, and the fact that this limitation exists at all is silly (I think all the limitations of custom element registries are silly, but that's a matter for another time), but the single global registry is currently the status quo. The linked article justifies user-customized tag names by saying

> A reusable element needs to allow registering itself with a different tag name.

But I think the flexibility the author desires simply isn't possible with a single global registry. Customizing the tag name solves one problem but creates another, equivalent problem.
